src/Generating/RandomizeStrategy.cpp

```diff
@@ -0,0 +1,264 @@
+
+// RandomizeStrategy.cpp
+
+// Implements the various classes descending from cPiece::cRandomizeStrategy
+
+#include "Globals.h"
+#include "RandomizeStrategy.h"
+#include "../Noise/Noise.h"
+#include "../Item.h"
+
+
+
+
+
+// Constant that is added to random seed
+static const int SEED_OFFSET = 135 * 13;
+
+
+
+
+
+// Emit a warning if the first param is true
+#define CONDWARNING(ShouldLog, Fmt, ...) \
+	do { \
+		if (ShouldLog) \
+		{ \
+			LOGWARNING(Fmt, __VA_ARGS__); \
+		} \
+	} while (false)
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+/** A randomize strategy that pseudo-randomly replaces wheat to various crops of different growth stages. */
+class cRandomizeStrategyCrops:
+	public cPiece::cRandomizeStrategy
+{
+public:
+	cRandomizeStrategyCrops(void) :
+		m_Seed()
+	{
+	}
+
+	virtual bool InitializeFromString(const AString & a_Params, bool a_LogWarnings) override
+	{
+		AString Params = a_Params;
+
+		auto idxPipe = Params.find('|');
+		if (idxPipe != AString::npos)
+		{
+			auto blockToReplace = cItem();
+			AString blockToReplaceStr = Params.substr(0, idxPipe);
+			StringToItem(blockToReplaceStr, blockToReplace);
+			if (blockToReplace.IsEmpty())
+			{
+				m_BlockToReplace = blockToReplace.m_ItemType;
+			}
+			else
+			{
+				// Failed to parse block to replace:
+				CONDWARNING(a_LogWarnings, "Cannot parse block to replace from string \"%s\"!", blockToReplaceStr.c_str());
+			}
+
+			Params = Params.substr(idxPipe + 1, Params.length() - 1);
+		}
+		else
+		{
+			m_BlockToReplace = E_BLOCK_WHEAT;
+		}
+
+		auto idxSqBracketEnd = Params.find(']');
+		auto idxSqBracketStart = Params.find('[');
+		if ((idxSqBracketEnd == Params.length()-1) && (idxSqBracketStart != AString::npos))
+		{
+			// Growth options defined
+
+			AString growthParamsStr = Params.substr(idxSqBracketStart + 1, Params.length()-idxSqBracketStart-2);
+			auto GrowthParams = StringSplitAndTrim(growthParamsStr, ",");
+
+			for (size_t i = 0; i < GrowthParams.size(); i++)
+			{
+				int GrowthValue;
+				if (!StringToInteger(GrowthParams[i], GrowthValue))
+				{
+					// Failed to parse the crop weight:
+					CONDWARNING(a_LogWarnings, "Cannot parse crops growth param from string \"%s\"!", GrowthParams[i].c_str());
+					return false;
+				}
+
+				switch (i)
+				{
+				case 0:
+					m_MinGrowthStage = GrowthValue;
+					break;
+				case 1:
+					m_MaxGrowthStage = GrowthValue;
+					break;
+				case 2:
+					m_MaxNoiseGrowthStage = GrowthValue;
+					break;
+				case 3:
+					m_MinNoiseGrowthStage = GrowthValue;
+					break;
+				default:
+					CONDWARNING(a_LogWarnings,"Unsupported crops growth param \"%d\"!", GrowthValue);
```

Comment:
Perhaps this should return false as well?

Reply:
Yeah perhaps, I intentionally didn't return there 😀